velox/functions/lib/ArrayFunctions.h

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright (c) Facebook, Inc. and its affiliates.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+#pragma once
+
+#include "velox/functions/Macros.h"
+
+namespace facebook::velox::functions {
+
+/// This class implements the remove_nulls function of presto and array_compact
+/// function of spark.
+///
+/// DEFINITION:
+/// presto: remove_nulls(array(E)) -> array(E)
+/// spark: array_compact(array(E)) -> array(E)
+/// Remove all null elements from array. Returns NULL as result if array is
+/// NULL.
+template <typename T>
+struct ArrayRemoveNullFunction {
+  VELOX_DEFINE_FUNCTION_TYPES(T);
+
+  // Fast path for primitives.
+  template <typename Out, typename In>
+  FOLLY_ALWAYS_INLINE void call(Out& out, const In& inputArray) {
+    if (inputArray.mayHaveNulls()) {
+      for (const auto& value : inputArray.skipNulls()) {
+        out.push_back(value);
+      }
+      return;
+    }
+
+    // No nulls, skip reading nullity.
+    for (const auto& value : inputArray) {
+      out.push_back(value);
+    }
+  }
+
+  // Generic implementation.
+  FOLLY_ALWAYS_INLINE void call(
+      out_type<Array<Generic<T1>>>& out,
+      const arg_type<Array<Generic<T1>>>& inputArray) {
+    if (inputArray.mayHaveNulls()) {
+      for (const auto& value : inputArray.skipNulls()) {
+        out.push_back(value);
+      }
+      return;
+    }
+
+    // No nulls, skip reading nullity.
+    for (const auto& value : inputArray) {
+      out.push_back(value);
+    }
+  }
+};
+
+/// This class implements the String version of the ArrayRemoveNullFunction.
+/// It optimizes performance by avoiding the copying of strings.
```

Comment:
Comment suggestion:

> /// Removes all NULL elements from the input string array. Returns empty string array if the
/// input string array is empty or all elements in it are NULL. Optimised by avoiding copy of strings.

Reply:
updated